back/engines/commercial/insights/app/controllers/insights/web_api/v1/views_controller.rb

```diff
@@ -13,13 +15,8 @@ def index
       end
 
       def create
-        view = authorize(Insights::View.new(create_params))
-        if view.save
-          CreateTnaTasksJob.perform_later(view)
-          topic_import_service.copy_assignments(view, current_user)
-          processed_service.set_processed(view.scope.ideas, [view.id])
-          #[TODO] feature-flag to only detect for premium
-          Insights::DetectCategoriesJob.perform_later(view)
+        view = Insights::Views::CreateService.new(current_user, create_params).execute
```

Comment:
It's really nice you've moved all this logic to the service. It makes this controller much cleaner and thinner :+1: 

But I'm not that sure about moving `authorize` to the service. 
1. I really like everything related to authentication and authorization to happen right in the controller. First, I can easily find it. Second, in my services, I can safely ignore everything related to auth. E.g. if somebody refactors `Views::CreateService`, they can forget about this exception and implement `Views::CreateService` without auth.
2. It seems we're not going to use `Views::CreateService` in other actions.
3. It breaks the rules (it seems we always use authorization in controllers). Maybe often it's better to make smth not perfect, but by rules. I'm afraid this `skip_after_action` can cause bugs in the future.

As an alternative, that moves all the code to service except `authorize` I can propose this code

```rb
view = Insights::Views::CreateService.new(current_user, create_params).execute { |view| authorize(view) }
```

Reply:
Thanks for the review! 🙏 

I see your point. I can agree that this solution is not perfect either, but I still think it's a better option in this case because:
- It avoids polluting the controller with additional logic.
- It makes the creation process easier to test (today, we don't have controller tests, we rely only on acceptance tests for that).
- It avoids repetition between controllers. That is not the case here, I grant you that, but I had the case very recently so I'm kind of sensitized to the issue. And I thought, it was a simple case to try out something new.

For info, I drew some inspiration from GitLab codebase here: https://docs.gitlab.com/ee/development/permissions.html#where-should-permissions-be-checked

> It breaks the rules (it seems we always use authorization in controllers).

Since it's an "experiment", it has to break some rules 🙂 And maybe, who knows, it will our new default in a few months.

> I'm afraid this skip_after_action can cause bugs in the future.

Why? `skip_after_action` is something we already use in quite a few places.

So unless it's a big NO-GO for you, I'd like to stick with that version.

Reply:
> It avoids polluting the controller with additional logic.

Agree

> It makes the creation process easier to test (today, we don't have controller tests, we rely only on acceptance tests for that).

Agree. As an alternative, I like to write more tests for the controller https://github.com/CitizenLabDotCo/cl2-admin/pull/48/files#diff-b8c6180bd81567deff977870439ba58e9d3fe517839e6df62900766198e9407f
But it's definitely not mainstream, so I cannot insist on it.

> For info, I drew some inspiration from GitLab codebase

Interesting read, thanks for sharing.

> skip_after_action is something we already use in quite a few places.

I personally wouldn't do it. I would even add `before_action :authorize` to make authorization mandatory. And then allow it in Policy.

> So unless it's a big NO-GO for you, I'd like to stick with that version.

Yes, it's absolutely fine to merge.

Reply:
> > It avoids polluting the controller with additional logic.

> Agree

Maybe cannot agree for this specific case as it's just simple `{ authorize(view) }`. But generally I agree :slightly_smiling_face:

Reply:
> > It avoids polluting the controller with additional logic.
> Maybe cannot agree for this specific case as it's just simple { authorize(view) }. But generally I agree 🙂

Yep, but I was comparing it to the original situation:
```ruby
view = authorize(Insights::View.new(create_params))
if view.save
  CreateTnaTasksJob.perform_later(view)
  topic_import_service.copy_assignments(view, current_user)
  processed_service.set_processed(view.scope.ideas, [view.id])
  #[TODO] feature-flag to only detect for premium
  Insights::DetectCategoriesJob.perform_later(view)
```

I don't really like the block approach. I find it a bit convoluted and it would make the testing of `CreateService` much more complex.

Reply:
> I personally wouldn't do it. I would even add before_action :authorize to make authorization mandatory. And then allow it in Policy.

How would perform the `authorize` bf the resource exists as it's case for `create` actions?